File: frontend/src/utils/validation.ts
const phonePattern = /^\+?\d+$/;
const personNamePattern = /^[A-Za-z][A-Za-z' -]*$/;
const postalCodePattern = /^[A-Za-z0-9][A-Za-z0-9 -]{1,18}[A-Za-z0-9]$/;
const usZipPattern = /^\d{5}(?:-\d{4})?$/;
const skuPattern = /^[A-Z0-9][A-Z0-9._-]{1,79}$/;

export const PHONE_MESSAGE = "Phone number must contain 10 to 15 digits and may start with +.";
export const PASSWORD_MESSAGE =
  "Password must include uppercase, lowercase, number, and special character.";

export function trimmed(value: string | undefined | null) {
  return value?.trim() ?? "";
}

export function optionalText(value: string | undefined | null) {
  const clean = trimmed(value);
  return clean || undefined;
}

export function validateRequiredText(value: string, label: string) {
  return trimmed(value) ? null : `${label} is required.`;
}

export function validatePersonName(value: string, label: string) {
  const clean = trimmed(value);
  if (!clean) return `${label} is required.`;
  if (!personNamePattern.test(clean)) {
    return `${label} may contain letters, spaces, hyphens, and apostrophes only.`;
  }
  return null;
}

export function validatePhone(value: string | undefined | null) {
  const clean = trimmed(value);
  if (!clean) return null;
  const digitCount = clean.startsWith("+") ? clean.length - 1 : clean.length;
  if (!phonePattern.test(clean) || digitCount < 10 || digitCount > 15) {
    return PHONE_MESSAGE;
  }
  return null;
}

export function validatePassword(value: string) {
  if (value.length < 8) return "Password must be at least 8 characters.";
  if (!/[A-Z]/.test(value) || !/[a-z]/.test(value) || !/\d/.test(value) || !/[^A-Za-z0-9]/.test(value)) {
    return PASSWORD_MESSAGE;
  }
  return null;
}

export function validatePositiveDecimal(value: string, label: string) {
  const clean = trimmed(value);
  const numberValue = Number(clean);
  if (!clean || Number.isNaN(numberValue)) return `${label} must be a valid number.`;
  if (!Number.isFinite(numberValue) || numberValue <= 0) return `${label} must be greater than 0.`;
  return null;
}

export function validateNonNegativeDecimal(value: string, label: string) {
  const clean = trimmed(value);
  const numberValue = Number(clean);
  if (!clean || Number.isNaN(numberValue)) return `${label} must be a valid number.`;
  if (!Number.isFinite(numberValue) || numberValue < 0) return `${label} cannot be negative.`;
  return null;
}

export function validateNonNegativeInteger(value: string, label: string) {
  const clean = trimmed(value);
  const numberValue = Number(clean);
  if (!clean || !Number.isInteger(numberValue)) return `${label} must be a whole number.`;
  if (numberValue < 0) return `${label} cannot be negative.`;
  return null;
}

export function validatePositiveInteger(value: string, label: string) {
  const integerError = validateNonNegativeInteger(value, label);
  if (integerError) return integerError;
  if (Number(trimmed(value)) < 1) return `${label} must be at least 1.`;
  return null;
}

export function validatePostalCode(value: string, country = "US") {
  const clean = trimmed(value);
  if (!clean) return "Postal code is required.";
  if (country.trim().toUpperCase() === "US") {
    return usZipPattern.test(clean) ? null : "US ZIP code must be 5 digits or ZIP+4 format.";
  }
  return postalCodePattern.test(clean) ? null : "Postal code contains invalid characters.";
}

export function validateSku(value: string) {
  const clean = trimmed(value).toUpperCase();
  if (!clean) return null;
  return skuPattern.test(clean)
    ? null
    : "SKU may contain uppercase letters, numbers, dots, hyphens, and underscores.";
}

export function validateOptionalHttpUrl(value: string, label: string) {
  const clean = trimmed(value);
  if (!clean) return null;
  if (!clean.startsWith("http://") && !clean.startsWith("https://")) {
    return `${label} must start with http:// or https://.`;
  }
  if (/\s/.test(clean)) return `${label} cannot contain whitespace.`;
  try {
    new URL(clean);
  } catch {
    return `${label} must be a valid URL.`;
  }
  return null;
}

export function validateCode(value: string, label: string) {
  const clean = trimmed(value).toUpperCase();
  if (!clean) return `${label} is required.`;
  return skuPattern.test(clean)
    ? null
    : `${label} may contain uppercase letters, numbers, dots, hyphens, and underscores.`;
}

export function firstError(...errors: Array<string | null>) {
  return errors.find((error): error is string => Boolean(error)) ?? null;
}

File: backend/app/schemas/validation.py
import re
from typing import Any

from app.core.config import get_settings

CONTROL_CHAR_PATTERN = re.compile(r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]")
SCRIPT_TAG_PATTERN = re.compile(r"<\s*/?\s*script\b", re.IGNORECASE)
PERSON_NAME_PATTERN = re.compile(r"^[A-Za-z][A-Za-z' -]*$")
PHONE_PATTERN = re.compile(r"^\+?\d+$")
POSTAL_CODE_PATTERN = re.compile(r"^[A-Za-z0-9][A-Za-z0-9 -]{1,18}[A-Za-z0-9]$")
US_ZIP_PATTERN = re.compile(r"^\d{5}(?:-\d{4})?$")
SKU_PATTERN = re.compile(r"^[A-Z0-9][A-Z0-9._-]{1,79}$")
CODE_PATTERN = re.compile(r"^[A-Z0-9][A-Z0-9._-]{1,79}$")


def clean_text(value: Any, *, field_name: str, required: bool = True) -> str | None:
    if value is None:
        if required:
            raise ValueError(f"{field_name} is required")
        return None
    if not isinstance(value, str):
        raise ValueError(f"{field_name} must be text")
    cleaned = value.strip()
    if not cleaned:
        if required:
            raise ValueError(f"{field_name} is required")
        return None
    if CONTROL_CHAR_PATTERN.search(cleaned) or SCRIPT_TAG_PATTERN.search(cleaned):
        raise ValueError(f"{field_name} contains unsupported characters")
    return cleaned


def clean_optional_text(value: Any, *, field_name: str) -> str | None:
    return clean_text(value, field_name=field_name, required=False)


def clean_optional_required_text(value: Any, *, field_name: str) -> str | None:
    if value is None:
        return None
    return clean_text(value, field_name=field_name)


def validate_person_name(value: Any, *, field_name: str) -> str:
    cleaned = clean_text(value, field_name=field_name)
    if cleaned is None or not PERSON_NAME_PATTERN.fullmatch(cleaned):
        raise ValueError(f"{field_name} may contain letters, spaces, hyphens, and apostrophes only")
    return cleaned


def validate_optional_person_name(value: Any, *, field_name: str) -> str | None:
    if value is None:
        return None
    return validate_person_name(value, field_name=field_name)


def validate_password_strength(value: Any) -> str:
    password = clean_text(value, field_name="Password")
    if password is None:
        raise ValueError("Password is required")
    if len(password) < 8:
        raise ValueError("Password must be at least 8 characters")
    if not re.search(r"[A-Z]", password):
        raise ValueError("Password must include an uppercase letter")
    if not re.search(r"[a-z]", password):
        raise ValueError("Password must include a lowercase letter")
    if not re.search(r"\d", password):
        raise ValueError("Password must include a number")
    if not re.search(r"[^A-Za-z0-9]", password):
        raise ValueError("Password must include a special character")
    return password


def validate_phone_number(value: Any) -> str | None:
    cleaned = clean_optional_text(value, field_name="Phone number")
    if cleaned is None:
        return None
    settings = get_settings()
    digit_count = len(cleaned[1:] if cleaned.startswith("+") else cleaned)
    if not PHONE_PATTERN.fullmatch(cleaned):
        raise ValueError("Phone number must contain digits only and may start with +")
    if digit_count < settings.phone_min_digits or digit_count > settings.phone_max_digits:
        raise ValueError(
            "Phone number must contain "
            f"{settings.phone_min_digits} to {settings.phone_max_digits} digits",
        )
    return cleaned


def validate_postal_code(value: Any, *, country: str | None = None) -> str:
    cleaned = clean_text(value, field_name="Postal code")
    if cleaned is None:
        raise ValueError("Postal code is required")
    if (country or "").strip().upper() == "US":
        if not US_ZIP_PATTERN.fullmatch(cleaned):
            raise ValueError("US ZIP code must be 5 digits or ZIP+4 format")
        return cleaned
    if not POSTAL_CODE_PATTERN.fullmatch(cleaned):
        raise ValueError("Postal code contains invalid characters")
    return cleaned


def validate_sku(value: Any) -> str:
    cleaned = clean_text(value, field_name="SKU")
    if cleaned is None:
        raise ValueError("SKU is required")
    normalized = cleaned.upper()
    if not SKU_PATTERN.fullmatch(normalized):
        raise ValueError(
            "SKU may contain uppercase letters, numbers, dots, hyphens, and underscores",
        )
    return normalized


def validate_code(value: Any, *, field_name: str) -> str:
    cleaned = clean_text(value, field_name=field_name)
    if cleaned is None:
        raise ValueError(f"{field_name} is required")
    normalized = cleaned.upper()
    if not CODE_PATTERN.fullmatch(normalized):
        raise ValueError(
            f"{field_name} may contain letters, numbers, dots, hyphens, and underscores",
        )
    return normalized
